src/persist-client/src/internal/encoding.rs

```diff
@@ -1607,4 +1614,48 @@ mod tests {
 
         proptest!(|(state in any_state::<u64>(0..3))| testcase(state));
     }
+
+    #[mz_ore::test]
+    fn check_data_versions() {
+        #[track_caller]
+        fn testcase(code: &str, data: &str, expected: Result<(), ()>) {
+            let code = Version::parse(code).unwrap();
+            let data = Version::parse(data).unwrap();
+            #[allow(clippy::disallowed_methods)]
+            let actual =
+                std::panic::catch_unwind(|| check_data_version(&code, &data)).map_err(|_| ());
+            assert_eq!(actual, expected);
+        }
+
+        testcase("0.10.0-dev", "0.10.0-dev", Ok(()));
+        testcase("0.10.0-dev", "0.10.0", Ok(()));
+        testcase("0.10.0-dev", "0.11.0-dev", Ok(()));
+        testcase("0.10.0-dev", "0.11.0", Ok(()));
+        testcase("0.10.0-dev", "0.12.0-dev", Err(()));
+        testcase("0.10.0-dev", "0.12.0", Err(()));
+        testcase("0.10.0-dev", "0.13.0-dev", Err(()));
+
+        testcase("0.10.0", "0.8.0-dev", Ok(()));
+        testcase("0.10.0", "0.8.0", Ok(()));
+        testcase("0.10.0", "0.9.0-dev", Ok(()));
+        testcase("0.10.0", "0.9.0", Ok(()));
+        testcase("0.10.0", "0.10.0-dev", Ok(()));
+        testcase("0.10.0", "0.10.0", Ok(()));
+        testcase("0.10.0", "0.11.0-dev", Ok(()));
+        testcase("0.10.0", "0.11.0", Ok(()));
+        testcase("0.10.0", "0.11.1", Ok(()));
+        testcase("0.10.0", "0.11.1000000", Ok(()));
+        testcase("0.10.0", "0.12.0-dev", Err(()));
+        testcase("0.10.0", "0.12.0", Err(()));
+        testcase("0.10.0", "0.13.0-dev", Err(()));
+
+        testcase("0.10.1", "0.11.0", Ok(()));
+        testcase("0.10.1", "0.11.1", Ok(()));
+        testcase("0.10.1", "0.11.100", Ok(()));
+
+        // This is probably a bad idea, but not much we can do, given the
+        // `state_version = max(code_version, prev_state_version)` logic we need
+        // to prevent rolling back an arbitrary number of versions.
```

Comment:
I don't totally follow why this is bad... using code-version 0.10.1 to read data-version 0.10.0 seems like a pretty common scenario. (eg. when we find an issue in a canary env.) Do you mean the other way around / am I missing something?